# Data and Time 

Python has the datetime module to help deal with timestamps in your code. Time values are represented with the time class. Times have attributes for hour, minute, second, and microsecond. They can also include time zone information. The arguments to initialize a time instance are optional, but the default of 0 is unlikely to be what you want.

## time
Lets take a look at how we can extract time information from the datetime module. We can create a time-stamp by specifying datetime.time(hour,minute,second,microsecond)

In [57]:
import datetime

t = datetime.time(4, 20, 1)
# Lets show the different compoenets
t

datetime.time(4, 20, 1)

In [58]:
time.gmtime(0)

time.struct_time(tm_year=1970, tm_mon=1, tm_mday=1, tm_hour=0, tm_min=0, tm_sec=0, tm_wday=3, tm_yday=1, tm_isdst=0)

In [2]:
print(t)

04:20:01


In [3]:
print('hour  :', t.hour)

hour  : 4


In [4]:
print('minute:', t.minute)

minute: 20


In [5]:
print('second:', t.second)

second: 1


In [6]:
print('microsecond:', t.microsecond)

microsecond: 0


In [7]:
print('tzinfo:', t.tzinfo)

tzinfo: None


Note: A time instance only holds values of time, and not a date associated with the time. 

We can also check the min and max values a time of day can have in the module:

In [8]:
print('Earliest  :', datetime.time.min)

Earliest  : 00:00:00


In [9]:
print('Latest    :', datetime.time.max)

Latest    : 23:59:59.999999


In [10]:
print('Resolution:', datetime.time.resolution)

Resolution: 0:00:00.000001


The min and max class attributes reflect the valid range of times in a single day.

## Dates
datetime (as you might suspect) also allows us to work with date timestamps. Calendar date values are represented with the date class. Instances have attributes for year, month, and day. It is easy to create a date representing today’s date using the today() class method.

Lets see some examples:

In [11]:
today = datetime.date.today()
print(today)

2020-04-27


In [12]:
print('ctime:', today.ctime())

ctime: Mon Apr 27 00:00:00 2020


In [13]:
print('tuple:', today.timetuple())

tuple: time.struct_time(tm_year=2020, tm_mon=4, tm_mday=27, tm_hour=0, tm_min=0, tm_sec=0, tm_wday=0, tm_yday=118, tm_isdst=-1)


In [14]:
print('ordinal:', today.toordinal())

ordinal: 737542


In [17]:
print('Year:', today.year)
print ('Mon :', today.month)
print ('Day :', today.day)

Year: 2020
Mon : 4
Day : 27


In [23]:
today.weekday()

0

As with time, the range of date values supported can be determined using the min and max attributes.

In [16]:
print ('Earliest  :', datetime.date.min)
print ('Latest    :', datetime.date.max)
print ('Resolution:', datetime.date.resolution)

Earliest  : 0001-01-01
Latest    : 9999-12-31
Resolution: 1 day, 0:00:00


Another way to create new date instances uses the replace() method of an existing date. For example, you can change the year, leaving the day and month alone.

In [24]:
d1 = datetime.date(2015, 3, 11)
print('d1:', d1)

d1: 2015-03-11


In [29]:
d2 = d1.replace(year=2010)
print('d2:', d2)

d2: 2010-03-11


In [30]:
import time

print("time.altzone %d " % time.altzone)

time.altzone -23400 


In [32]:
t = time.localtime()
print("time.asctime(t): %s " % time.asctime(t))

time.asctime(t): Mon Apr 27 19:13:15 2020 


In [34]:
time.clock() # CPU Clock

84.0295481

In [37]:
print("time.localtime() : s" ,time.localtime())

time.localtime() : s time.struct_time(tm_year=2020, tm_mon=4, tm_mday=27, tm_hour=19, tm_min=16, tm_sec=20, tm_wday=0, tm_yday=118, tm_isdst=0)


In [47]:
struct_time = time.strptime("30 Nov 00", "%d %b %y")
print("returned tuple: s ",struct_time)

returned tuple: s  time.struct_time(tm_year=2000, tm_mon=11, tm_mday=30, tm_hour=0, tm_min=0, tm_sec=0, tm_wday=3, tm_yday=335, tm_isdst=-1)


In [42]:
from datetime import datetime
datetime(year=2020, month=4, day=14)

datetime.datetime(2020, 4, 14, 0, 0)

In [43]:
from dateutil import parser

In [49]:
date =parser.parse('27th of July 2020')

In [52]:
date.strftime('%a') # Weekday as locale’s abbreviated name.

'Mon'

In [54]:
date.strftime('%A') # Weekday as locale’s full name

'Monday'

In [56]:
date.strftime('%w') # Weekday as a decimal number, where 0 is Sunday and 6 is Saturday.

'1'

<div class="section" id="strftime-and-strptime-behavior">
<span id="strftime-strptime-behavior"></span><span id="index-0"></span><h2><code class="xref py py-meth docutils literal notranslate"><span class="pre">strftime()</span></code> and <code class="xref py py-meth docutils literal notranslate"><span class="pre">strptime()</span></code> Behavior<a class="headerlink" href="#strftime-and-strptime-behavior" title="Permalink to this headline">¶</a></h2>
<p><a class="reference internal" href="#datetime.date" title="datetime.date"><code class="xref py py-class docutils literal notranslate"><span class="pre">date</span></code></a>, <a class="reference internal" href="#datetime.datetime" title="datetime.datetime"><code class="xref py py-class docutils literal notranslate"><span class="pre">datetime</span></code></a>, and <a class="reference internal" href="#datetime.time" title="datetime.time"><code class="xref py py-class docutils literal notranslate"><span class="pre">time</span></code></a> objects all support a
<code class="docutils literal notranslate"><span class="pre">strftime(format)</span></code> method, to create a string representing the time under the
control of an explicit format string.</p>
<p>Conversely, the <a class="reference internal" href="#datetime.datetime.strptime" title="datetime.datetime.strptime"><code class="xref py py-meth docutils literal notranslate"><span class="pre">datetime.strptime()</span></code></a> class method creates a
<a class="reference internal" href="#datetime.datetime" title="datetime.datetime"><code class="xref py py-class docutils literal notranslate"><span class="pre">datetime</span></code></a> object from a string representing a date and time and a
corresponding format string.</p>
<p>The table below provides a high-level comparison of <code class="xref py py-meth docutils literal notranslate"><span class="pre">strftime()</span></code>
versus <code class="xref py py-meth docutils literal notranslate"><span class="pre">strptime()</span></code>:</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 11%">
<col style="width: 37%">
<col style="width: 52%">
</colgroup>
<thead>
<tr class="row-odd"><th class="head"></th>
<th class="head"><p><code class="docutils literal notranslate"><span class="pre">strftime</span></code></p></th>
<th class="head"><p><code class="docutils literal notranslate"><span class="pre">strptime</span></code></p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Usage</p></td>
<td><p>Convert object to a string according to a given format</p></td>
<td><p>Parse a string into a <a class="reference internal" href="#datetime.datetime" title="datetime.datetime"><code class="xref py py-class docutils literal notranslate"><span class="pre">datetime</span></code></a> object given a corresponding format</p></td>
</tr>
<tr class="row-odd"><td><p>Type of method</p></td>
<td><p>Instance method</p></td>
<td><p>Class method</p></td>
</tr>
<tr class="row-even"><td><p>Method of</p></td>
<td><p><a class="reference internal" href="#datetime.date" title="datetime.date"><code class="xref py py-class docutils literal notranslate"><span class="pre">date</span></code></a>; <a class="reference internal" href="#datetime.datetime" title="datetime.datetime"><code class="xref py py-class docutils literal notranslate"><span class="pre">datetime</span></code></a>; <a class="reference internal" href="#datetime.time" title="datetime.time"><code class="xref py py-class docutils literal notranslate"><span class="pre">time</span></code></a></p></td>
<td><p><a class="reference internal" href="#datetime.datetime" title="datetime.datetime"><code class="xref py py-class docutils literal notranslate"><span class="pre">datetime</span></code></a></p></td>
</tr>
<tr class="row-odd"><td><p>Signature</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">strftime(format)</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">strptime(date_string,</span> <span class="pre">format)</span></code></p></td>
</tr>
</tbody>
</table>
<div class="section" id="strftime-and-strptime-format-codes">
<h3><code class="xref py py-meth docutils literal notranslate"><span class="pre">strftime()</span></code> and <code class="xref py py-meth docutils literal notranslate"><span class="pre">strptime()</span></code> Format Codes<a class="headerlink" href="#strftime-and-strptime-format-codes" title="Permalink to this headline">¶</a></h3>
<p>The following is a list of all the format codes that the 1989 C standard
requires, and these work on all platforms with a standard C implementation.</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 15%">
<col style="width: 43%">
<col style="width: 32%">
<col style="width: 9%">
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Directive</p></th>
<th class="head"><p>Meaning</p></th>
<th class="head"><p>Example</p></th>
<th class="head"><p>Notes</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">%a</span></code></p></td>
<td><p>Weekday as locale’s
abbreviated name.</p></td>
<td><div class="line-block">
<div class="line">Sun, Mon, …, Sat
(en_US);</div>
<div class="line">So, Mo, …, Sa
(de_DE)</div>
</div>
</td>
<td><p>(1)</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">%A</span></code></p></td>
<td><p>Weekday as locale’s full name.</p></td>
<td><div class="line-block">
<div class="line">Sunday, Monday, …,
Saturday (en_US);</div>
<div class="line">Sonntag, Montag, …,
Samstag (de_DE)</div>
</div>
</td>
<td><p>(1)</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">%w</span></code></p></td>
<td><p>Weekday as a decimal number,
where 0 is Sunday and 6 is
Saturday.</p></td>
<td><p>0, 1, …, 6</p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">%d</span></code></p></td>
<td><p>Day of the month as a
zero-padded decimal number.</p></td>
<td><p>01, 02, …, 31</p></td>
<td><p>(9)</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">%b</span></code></p></td>
<td><p>Month as locale’s abbreviated
name.</p></td>
<td><div class="line-block">
<div class="line">Jan, Feb, …, Dec
(en_US);</div>
<div class="line">Jan, Feb, …, Dez
(de_DE)</div>
</div>
</td>
<td><p>(1)</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">%B</span></code></p></td>
<td><p>Month as locale’s full name.</p></td>
<td><div class="line-block">
<div class="line">January, February,
…, December (en_US);</div>
<div class="line">Januar, Februar, …,
Dezember (de_DE)</div>
</div>
</td>
<td><p>(1)</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">%m</span></code></p></td>
<td><p>Month as a zero-padded
decimal number.</p></td>
<td><p>01, 02, …, 12</p></td>
<td><p>(9)</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">%y</span></code></p></td>
<td><p>Year without century as a
zero-padded decimal number.</p></td>
<td><p>00, 01, …, 99</p></td>
<td><p>(9)</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">%Y</span></code></p></td>
<td><p>Year with century as a decimal
number.</p></td>
<td><p>0001, 0002, …, 2013,
2014, …, 9998, 9999</p></td>
<td><p>(2)</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">%H</span></code></p></td>
<td><p>Hour (24-hour clock) as a
zero-padded decimal number.</p></td>
<td><p>00, 01, …, 23</p></td>
<td><p>(9)</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">%I</span></code></p></td>
<td><p>Hour (12-hour clock) as a
zero-padded decimal number.</p></td>
<td><p>01, 02, …, 12</p></td>
<td><p>(9)</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">%p</span></code></p></td>
<td><p>Locale’s equivalent of either
AM or PM.</p></td>
<td><div class="line-block">
<div class="line">AM, PM (en_US);</div>
<div class="line">am, pm (de_DE)</div>
</div>
</td>
<td><p>(1),
(3)</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">%M</span></code></p></td>
<td><p>Minute as a zero-padded
decimal number.</p></td>
<td><p>00, 01, …, 59</p></td>
<td><p>(9)</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">%S</span></code></p></td>
<td><p>Second as a zero-padded
decimal number.</p></td>
<td><p>00, 01, …, 59</p></td>
<td><p>(4),
(9)</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">%f</span></code></p></td>
<td><p>Microsecond as a decimal
number, zero-padded on the
left.</p></td>
<td><p>000000, 000001, …,
999999</p></td>
<td><p>(5)</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">%z</span></code></p></td>
<td><p>UTC offset in the form
<code class="docutils literal notranslate"><span class="pre">±HHMM[SS[.ffffff]]</span></code> (empty
string if the object is
naive).</p></td>
<td><p>(empty), +0000,
-0400, +1030,
+063415,
-030712.345216</p></td>
<td><p>(6)</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">%Z</span></code></p></td>
<td><p>Time zone name (empty string
if the object is naive).</p></td>
<td><p>(empty), UTC, EST, CST</p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">%j</span></code></p></td>
<td><p>Day of the year as a
zero-padded decimal number.</p></td>
<td><p>001, 002, …, 366</p></td>
<td><p>(9)</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">%U</span></code></p></td>
<td><p>Week number of the year
(Sunday as the first day of
the week) as a zero padded
decimal number. All days in a
new year preceding the first
Sunday are considered to be in
week 0.</p></td>
<td><p>00, 01, …, 53</p></td>
<td><p>(7),
(9)</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">%W</span></code></p></td>
<td><p>Week number of the year
(Monday as the first day of
the week) as a decimal number.
All days in a new year
preceding the first Monday
are considered to be in
week 0.</p></td>
<td><p>00, 01, …, 53</p></td>
<td><p>(7),
(9)</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">%c</span></code></p></td>
<td><p>Locale’s appropriate date and
time representation.</p></td>
<td><div class="line-block">
<div class="line">Tue Aug 16 21:30:00
1988 (en_US);</div>
<div class="line">Di 16 Aug 21:30:00
1988 (de_DE)</div>
</div>
</td>
<td><p>(1)</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">%x</span></code></p></td>
<td><p>Locale’s appropriate date
representation.</p></td>
<td><div class="line-block">
<div class="line">08/16/88 (None);</div>
<div class="line">08/16/1988 (en_US);</div>
<div class="line">16.08.1988 (de_DE)</div>
</div>
</td>
<td><p>(1)</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">%X</span></code></p></td>
<td><p>Locale’s appropriate time
representation.</p></td>
<td><div class="line-block">
<div class="line">21:30:00 (en_US);</div>
<div class="line">21:30:00 (de_DE)</div>
</div>
</td>
<td><p>(1)</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">%%</span></code></p></td>
<td><p>A literal <code class="docutils literal notranslate"><span class="pre">'%'</span></code> character.</p></td>
<td><p>%</p></td>
<td></td>
</tr>
</tbody>
</table>
<p>Several additional directives not required by the C89 standard are included for
convenience. These parameters all correspond to ISO 8601 date values.</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 15%">
<col style="width: 43%">
<col style="width: 32%">
<col style="width: 9%">
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Directive</p></th>
<th class="head"><p>Meaning</p></th>
<th class="head"><p>Example</p></th>
<th class="head"><p>Notes</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">%G</span></code></p></td>
<td><p>ISO 8601 year with century
representing the year that
contains the greater part of
the ISO week (<code class="docutils literal notranslate"><span class="pre">%V</span></code>).</p></td>
<td><p>0001, 0002, …, 2013,
2014, …, 9998, 9999</p></td>
<td><p>(8)</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">%u</span></code></p></td>
<td><p>ISO 8601 weekday as a decimal
number where 1 is Monday.</p></td>
<td><p>1, 2, …, 7</p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">%V</span></code></p></td>
<td><p>ISO 8601 week as a decimal
number with Monday as
the first day of the week.
Week 01 is the week containing
Jan 4.</p></td>
<td><p>01, 02, …, 53</p></td>
<td><p>(8),
(9)</p></td>
</tr>
</tbody>
</table>
<p>These may not be available on all platforms when used with the <code class="xref py py-meth docutils literal notranslate"><span class="pre">strftime()</span></code>
method. The ISO 8601 year and ISO 8601 week directives are not interchangeable
with the year and week number directives above. Calling <code class="xref py py-meth docutils literal notranslate"><span class="pre">strptime()</span></code> with
incomplete or ambiguous ISO 8601 directives will raise a <a class="reference internal" href="exceptions.html#ValueError" title="ValueError"><code class="xref py py-exc docutils literal notranslate"><span class="pre">ValueError</span></code></a>.</p>
<p>The full set of format codes supported varies across platforms, because Python
calls the platform C library’s <code class="xref py py-func docutils literal notranslate"><span class="pre">strftime()</span></code> function, and platform
variations are common. To see the full set of format codes supported on your
platform, consult the <em class="manpage"><a class="manpage reference external" href="https://manpages.debian.org/strftime(3)">strftime(3)</a></em> documentation.</p>
<div class="versionadded">
<p><span class="versionmodified added">New in version 3.6: </span><code class="docutils literal notranslate"><span class="pre">%G</span></code>, <code class="docutils literal notranslate"><span class="pre">%u</span></code> and <code class="docutils literal notranslate"><span class="pre">%V</span></code> were added.</p>
</div>
</div>
</div>

# Arithmetic
We can perform arithmetic on date objects to check for time differences. For example:

In [27]:
d1-d2

datetime.timedelta(1826)

This give us the difference in days between the two dates. You can use the timedelta method to specify various units of times (day,minutes,hours,etc...)

Great! You should now have a basic understanding of how to use datetime with Python to work with timestamps in your code!